src/node_modules/app/elasticsearch/index.js

```diff
@@ -25,6 +37,83 @@ export function getESType (field) {
 	return 'object';
 }
 
+const revealTypeWith = defaults => _.pipe([
+	_.pull(defaults),
+	_.when(hasIterableLength1, _.head)
+]);
+const revealType = revealTypeWith(['undefined']);
+export const extractBaseType = _.when(isArray, revealType);
+
+const isStringLiteral = _.allOf([isString, makeStartsWith('"')]);
+export const isStringLiteralUnion = _.allOf([
+	isArray,
+	_.every(_.anyOf([
+		_.is('undefined'),
+		isStringLiteral
+	]))
+]);
+
+// export const isArrayType = type => '__array' in type; // TODO: new test
+export const isArrayType = makeEndsWith('[]');
+export const isJsonType = type => type === 'json';
+export const isRecordType = type => '__record' in type;
+
+
+function getDisplayType (fullType, indentation = 0) {
+	if (isString(fullType)) {
+		return fullType;
+	}
+	if (isArray(fullType)) {
+		return fullType
+		.filter(type => type !== 'undefined')
+		.map(type => getDisplayType(type, indentation + 1))
+		.join(' | ');
+	}
+	if (isObject(fullType)) {
+		if ('__array' in fullType) {
+			return `${getDisplayType(fullType.__array, indentation + 1)}[]`;
+		}
+		if ('__legacyName' in fullType) {
+			return fullType.__legacyName;
+		}
+		const paramNames = Object.keys(fullType);
+		const result = paramNames
+		.map(name => {
+			return `${'\t'.repeat(indentation)}${name}: ${getDisplayType(fullType[name], indentation + 1)};`;
+		})
+		.join('\n');
+		return ['{\n', result, '\n', '\t'.repeat(indentation >= 1 ? indentation - 1 : 0), '}'].join('');
+	}
+	throw new Error('Failed to parse type data, bad type definition.');
+}
```

Comment:
Suggesting some simplifications, probably we can do much more here:

```suggestion
export const hasArray = _.has('__array');
export const hasLegacyName = _.has('__legacyName');
export const equalsUndefined = _.is('undefined');
export const tab = indentation => _.partial(_.repeat, ['\t', indentation]));

function getDisplayType (fullType, indentation = 0) {
	if (isString(fullType)) {
		return fullType;
	}

	const indentDisplayType = _.partial(getDisplayType, [_, indentation + 1]);

	if (isArray(fullType)) {
		return fullType.filter(equalsUndefined).map(indentDisplayType).join(' | ');
	}
	if (isObject(fullType)) {
		if (hasArray(fullType)) {
			return `${indentDisplayType(fullType.__array)}[]`;
		}
		if (hasLegacyName(fullType)) {
			return fullType.__legacyName;
		}

		const pairs = _.pairs(fullType).map(([key, value]) => {
			return `${tab(indentation)}${key}: ${indentDisplayType(value)};`;
		});
		pairs.join('\n');

		return [
			'{',
			pairs,
			`${(tab(indentation >= 1 ? indentation - 1 : 0))}}`,
		].join('\n');
	}

	return fullType;

	// throw new Error('Failed to parse type data, bad type definition.');
}
```